File: languages/r/ruby/lambda-core.rb
# LOGIC
_true = ->(x) { ->(y) { x } }
_false = ->(x) { ->(y) { y } }
_not = ->(b) { b[_false][_true] }
_and = ->(b1) { ->(b2) { b1[b2][_false] } }
_or = ->(b1) { ->(b2) { b1[_true][b2] } }

# CHURCH NUMERALS
_zero = ->(f) { ->(x) { x } }
_succ = ->(n) { ->(f) { ->(x) { f[n[f][x]] } } }
_pred = ->(n) {
  ->(f) {
    ->(x) {
      n[->(g) { ->(h) { h[g[f]] } }][->(u) { x }][->(a) { a }]
    }
  }
}
_one = _succ[_zero]

# HELPERS - not pure lambda calculus
def read_bool(b)
  puts b["t"]["f"]
end

def read_church(n)
  puts n[->(x) { x + 1 }][0]
end

# EXAMPLES
puts "LOGIC"
puts "---------------"
puts "TRUE/FALSE"
read_bool(_true) # t
read_bool(_false) # f

puts "NOT"
read_bool(_not[_true]) # f
read_bool(_not[_false]) # t

puts "AND"
read_bool(_and[_true][_true]) # t
read_bool(_and[_true][_false]) # f
read_bool(_and[_false][_true]) # f
read_bool(_and[_false][_false]) # f

puts "OR"
read_bool(_or[_true][_true]) # t
read_bool(_or[_true][_false]) # t
read_bool(_or[_false][_true]) # t
read_bool(_or[_false][_false]) # f

puts "\nCHURCH NUMERALS"
puts "---------------"
puts "ZERO/SUCC"
read_church(_zero) # 0
read_church(_one) # 1
read_church(_succ[_one]) # 2
read_church(_succ[_succ[_one]]) # 3

puts "PRED"
read_church(_pred[_one]) # 0
read_church(_pred[_succ[_one]]) # 1
read_church(_pred[_succ[_succ[_one]]]) # 2

File: languages/c/c++/lambda-core.cpp
#include <print>

namespace logic {
constexpr auto _true = [](const auto &x) {
  return [&x](const auto &) { return x; };
};

constexpr auto _false = [](const auto &) {
  return [](const auto &y) { return y; };
};

constexpr auto _not = [](const auto &x) { return x(_false)(_true); };

constexpr auto _and = [](const auto &x) {
  return [&x](const auto &y) { return x(y)(_false); };
};

constexpr auto _or = [](const auto &x) {
  return [&x](const auto &y) { return x(_true)(y); };
};

namespace test {
constexpr bool convert(const auto &x) { return x(true)(false); }

void print(const auto &x) { std::println("{}", convert(x)); }
} // namespace test
} // namespace logic

namespace church {
constexpr auto zero = [](const auto &) {
  return [](const auto &x) { return x; };
};

constexpr auto succ = [](const auto &n) {
  return [&n](const auto &f) {
    return [&n, &f](const auto &x) { return f(n(f)(x)); };
  };
};

constexpr auto one = succ(zero);

constexpr auto pred = [](const auto &n) {
  return [&n](const auto &f) {
    return [&n, &f](auto x) {
      return n([&f](const auto &g) {
        return [&f, g](const auto &h) { return h(g(f)); };
      })([x](const auto &u) { return x; })([](const auto &a) { return a; });
    };
  };
};

namespace test {
constexpr unsigned int convert(const auto &n) {
  return n([](const auto &x) { return x + 1; })(0);
}

void print(const auto &n) { std::println("{}", convert(n)); }
} // namespace test
} // namespace church

int main() {
  using namespace logic;
  using namespace church;

  static_assert(logic::test::convert(_true) == true);
  static_assert(logic::test::convert(_false) == false);
  static_assert(logic::test::convert(_and(_true)(_true)) == true);
  static_assert(logic::test::convert(_and(_true)(_false)) == false);
  static_assert(logic::test::convert(_or(_true)(_true)) == true);
  static_assert(logic::test::convert(_or(_true)(_false)) == true);
  logic::test::print(_true);

  static_assert(church::test::convert(zero) == 0);
  static_assert(church::test::convert(succ(zero)) == 1);
  static_assert(church::test::convert(pred(succ(one))) == 1);
  church::test::print(one);
}
